@here/harp-mapview/lib/geometry/DisplacedMesh.ts

```diff
@@ -0,0 +1,96 @@
+/*
+ * Copyright (C) 2020 HERE Europe B.V.
+ * Licensed under Apache 2.0, see full license in LICENSE
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+import { hasDisplacementFeature } from "@here/harp-materials";
+import * as THREE from "three";
+import { DisplacedBufferAttribute } from "./DisplacedBufferAttribute";
+import { DisplacedBufferGeometry, DisplacementRange } from "./DisplacedBufferGeometry";
+
+/**
+ * Mesh with geometry modified by a displacement map. Overrides raycasting behaviour to apply
+ * displacement map before intersection test.
+ * @internal
+ */
+export class DisplacedMesh extends THREE.Mesh {
+    private static displacedPositions?: DisplacedBufferAttribute;
+
+    private static getDisplacedPositionAttribute(
+        geometry: THREE.BufferGeometry,
+        displacementMap: THREE.DataTexture
+    ): DisplacedBufferAttribute {
+        // Reuse same buffer attribute for all meshes since it's only needed during the
+        // intersection test.
+        if (!DisplacedMesh.displacedPositions) {
+            DisplacedMesh.displacedPositions = new DisplacedBufferAttribute(
+                geometry.attributes.position,
+                geometry.attributes.normal,
+                geometry.attributes.uv,
+                displacementMap
+            );
+        } else {
+            DisplacedMesh.displacedPositions.reset(
+                geometry.attributes.position,
+                geometry.attributes.normal,
+                geometry.attributes.uv,
+                displacementMap
+            );
+        }
+        return DisplacedMesh.displacedPositions;
+    }
+
+    m_displacedGeometry?: DisplacedBufferGeometry;
+
+    /**
+     * Creates an instance of displaced mesh.
+     * @param m_getDisplacementRange Displacement values range getter.
+     * @param [geometry] Original geometry to displace.
+     * @param [material] Material(s) to be used by the mesh. All must have the same displacement
+     * map.
+     */
+    constructor(
+        private m_getDisplacementRange: () => DisplacementRange,
+        geometry?: THREE.Geometry | THREE.BufferGeometry,
+        material?: THREE.Material | THREE.Material[]
```

Comment:
Why not check the displacement map to be exactly equal (ref) or have equal properties? This may reduce the potential errors.

Reply:
With what do you want to compare the displacement map?

Reply:
We could for example check if material has displacementMap defined, this is property appears for both build-on THREE.js materials and our feature mixins, keep in mind that displacementMap is Texture | null (not undefined). 
This is at least minimum requirement that gives the sense for using this class instead of standard mesh, otherwise it will be simply wrapper that works as any other mesh. 
Of course I see the test in raycast() method, but shouldn't we test before, if somebody is using the class with its designed purpose? I am leaving decision to you, but wanted to pinpoint that clearly.

Reply:
ok, I'm gonna add the checks as assertions both in the constructor and raycast functions.